### PR TITLE
Allow overriding types for other modules in prelude

### PR DESCRIPTION
**Pull request by zuiderkwast**
With this change, we're also not using the remote spec syntax
which is not supported in OTP anymore.

### CODE REVIEW

**Comment by zuiderkwast** [comment]
This is the follow-up to #191, as suggested by @gomoripeti in https://github.com/josefs/Gradualizer/issues/187#issuecomment-534583656.

I used the syntax `-override_module(Module)`, since it's not only specs anymore but also types and records. One known problem was fixed.

File: src/gradualizer_db.erl [hunk 1]
```diff
@@ -253,9 +253,18 @@ code_change(_OldVsn, State, _Extra) ->
...
+    FormsByModule = partition_forms_by_override_module(gradualizer_prelude, Forms),
+    %% Import forms each of the modules to override
+    State1 = lists:foldl(fun ({Module, FormsThisModule}, StateAcc) ->
+                                 import_absform(Module, FormsThisModule, StateAcc)
+                         end,
+                         State,
+                         FormsByModule),
+    %% Mark the just overridden modules as not yet loaded, to make sure they
+    %% are loaded on demand
+    State1#state{loaded = Loaded#{gradualizer_prelude => true}}.
```
Comment (gomoripeti) [inline src/gradualizer_db.erl]: hah, nice

**Comment by zuiderkwast** [comment]
That's a good point @josefs.

The `'module:name'` syntax would be a bit strange for overriding internal types though, e.g. `deep_list` below, a type which is not exported from lists.

```Erlang
-type 'lists:deep_list'(A) :: [A | deep_list(A)].
-spec 'lists:flatten'(deep_list(A))      -> [A].
```

Here, because of the `'lists:deep_list'` representation of the name, these definitions would be treated as part of the lists module, so in the RHS, simply `deep_lists` is used. Also in the spec below, `deep_lists` is used without module prefix. This looks a bit strange. The alternative would be to write `'lists:deep_list'` everywhere...

```Erlang
-type 'lists:deep_list'(A) :: [A | 'lists:deep_list'(A)].
-spec 'lists:flatten'('lists:deep_list'(A))      -> [A].
```

...but then we'd obviously have to handle this syntax in a lot more places.

The beauty of the `-override_type(lists)` syntax IMO is that it's clear that the code below this attribute is simply taken as part of the lists module, unchanged. (It was also easy to implement.)

----

An alternative would be to have one file per module we want to override, e.g. `priv/overrides/lists.erl`. I would put them under `priv` so that rebar doesn't think the're erl files to be compiled. Instead we can read these files in the prelude parse transform and let them be returned by the prelude module.

**Comment by josefs** [comment]
I like the idea of having one file per overridden module. We've already talked about splitting up this module anyway so that seems like a really nice solution actually!

**Comment by gomoripeti** [comment]
One more thing we could consider when defining the format of the override file is that in the future users could also define their own override files (eg for 3rd party libs). In that case I think its more convenient to allow a single file to override let's say 3 specs each from 3 different libs. Similar to a single dialyzer.ignore file. In that example it would be a bit less convenient to create 3 files for 3 modules each with only 1 function.

But I also agree the prelude file grew too large so it makes sense to split it up and support multiple override files.

**Comment by josefs** [comment]
@gomoripeti yes, if we allow users to override libraries in the future it'd be somewhat less convenient to have one file per overridden module. Are you arguing that this inconvenience is a killer and that we should allow multiple overridden modules per file? 
For me it doesn't weigh particularly heavy and I think the problems that are solved by having multiple files clearly wins out in this case.

**Comment by Zalastax** [comment]
Making a feature easy to explain is a high priority for me so I want us to go for split files. Users won't complain over having to split their preludes and split files is more intuitive.

**Comment by zuiderkwast** [comment]
We can allow users so specify their own *prelude directories*,

```sh
$ gradualizer --prelude-dir ~/my/prelude -pa xyz -- files_to_check`
```

Such dir can have the same structure, i.e. one file per module to override stuff for (e.g. lists.erl). Many small files? Who cares? It's a simple interface. Maybe we should consider a different suffix than `.erl` though, as these files won't compile as Erlang modules. How about `.erlspec` or similar?

**Comment by josefs** [comment]
Having a separate file suffix sounds like a good idea. I don't have a strong opinion on the actual name

File: src/gradualizer_prelude_parse_trans.erl [hunk 2]
```diff
@@ -3,13 +3,27 @@
...
+    [Form | replace_get_modules_and_forms(RestForms)].
+
+%% The value to be returned by the function in the parse transformed module
+-spec get_module_forms_tuples() -> [{module(), gradualizer_file_utils:abstract_forms()}].
+get_module_forms_tuples() ->
+    Files = filelib:wildcard(filename:join([code:priv_dir(gradualizer), "prelude", "*.erl"])),
+    lists:map(fun get_module_and_forms/1, Files).
+
+%% Parses and returns the forms of a file along with the module given in the -module attribute
+-spec get_module_and_forms(file:filename()) -> gradualizer_file_utils:abstract_forms().
```
Comment (josefs) [inline src/gradualizer_prelude_parse_trans.erl]: dialyzer seems to have a problem with the return type.

Comment (zuiderkwast) [reply src/gradualizer_prelude_parse_trans.erl]: Yes, it was a bad spec (AKA type error).

File: priv/prelude/mnesial.specs.erl [hunk 3]
```diff
@@ -0,0 +1,164 @@
+-module(mnesia).
```
Comment (KennethL) [inline priv/prelude/mnesial.specs.erl]: Why is the filename mnesial.specs.erl? Should it not be mnesia.specs.erl?

Comment (zuiderkwast) [reply priv/prelude/mnesial.specs.erl]: Oops, you're right of course! (The file name is not checked/used so it makes no functional difference but still.)

**Comment by gomoripeti** [comment]
great work @zuiderkwast 

Im not sure if this is the case but  if `gradualizer_prelude.beam` is already compiled and one changes a specs file, rebar might not recompile the module as `gradualizer_prelude.erl` hasn't changed and rebar does not know there is a dependency between the specs file and the source file. If this is indeed a problem, maybe better to use a compile hook escript instead of the parse trans to conditionally regenerate the `gradualizer_prelude.erl` file. What do you all think?

**Comment by Zalastax** [comment]
I don't trust the recompilation process anyway so I always delete the `_build` folder. I'm happy with the current solution.

**Comment by zuiderkwast** [comment]
@gomoripeti 
> Im not sure if this is the case but if `gradualizer_prelude.beam` is already compiled and one changes a specs file, rebar might not recompile the module as `gradualizer_prelude.erl` hasn't changed and rebar does not know there is a dependency between the specs file and the source file.

Right.... In a Makefile (let's say erlang.mk) it would be so easy to just add a dependency, like `ebin/gradualizer_prelude.beam: priv/prelude` or so. (Why did people invent rebar anyway?)

> If this is indeed a problem, maybe better to use a compile hook escript instead of the parse trans to conditionally regenerate the `gradualizer_prelude.erl` file. What do you all think?

I guess it needs to be addressed whenever someone adds or changes a prelude spec the next time. Otherwise users won't see any effect. Personally, I don't want to have hard dependencies on rebar, so I prefer a parse transform, but as long as rebar is used to build, a hook for at least invalidating the prelude would be good. (IMO gradualizer should be compatible with rebar as a plugin, etc. but preferably it should also be possible to build and use it with e.g. Mix or GNU Make or even Erlang `make:all()` as an alternative. It shouldn't be that hard since we don't have any dependencies. For dev stuff like running the tests, requiring rebar is fine.)

@Zalastax 
> I don't trust the recompilation process anyway so I always delete the _build folder. I'm happy with the current solution.

Wow, it really is a ridiculous build system then. IMO you can't expect people to `make clean` at every update.

**Comment by josefs** [comment]
I'm just going to add that I've lost a lot of time because the recompilation checker in rebar doesn't do the right thing. And I still have an issue when using dialyzer which I cannot seem to fix no matter how many files I delete. At this point I'm considering setting my laptop on fire to see if that helps.
I'm very tempted to replace rebar for the day-to-day development of gradualizer because it just slows me down right now. Though as @zuiderkwast says, we should still make sure we're compatible with rebar e.g. as a plugin. But we should have another build system which doesn't act as a roadblock.

**Pull request merged**